wallmatrix/sources/__init__.py

```diff
@@ -16,18 +21,18 @@ def teardown(self):
         "Tear down any lasting objects required by the source."
         pass
 
-    def get_data(self):
+    @abstractmethod
+    def get_data(self) -> T:
         """Return the data required by the source.
         Will be called once every minute.
         Useful to avoid hitting API quotas."""
-        pass
 
-    def get_image(self, data):
+    @abstractmethod
+    def get_image(self, data: T | None) -> Image:
```

Comment:
What's the justification for the `| None`? I think the data cache should be populated before `get_image` is called in every circumstance, although that may need to be refactored a bit for the type system to pick up on it.

Reply:
fixed by initializing it to non-null in the constructor